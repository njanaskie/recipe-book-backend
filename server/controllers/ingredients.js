const ingredientsRouter = require('express').Router();
const Ingredient = require('../models/ingredient')


getIngredients = async (req, res) => {
    const auth = req.currentUser;

    if (auth) {
        const ingredients = await Ingredient.find({}).sort( { name: 1 } )
        return res.json(ingredients.map((ingredient) => ingredient.toJSON() ));
    }
    return res.status(403).send('Not authorized get')
}
// ingredientsRouter.get('/ingredients', (req, res) => {
//     // const auth = req.currentUser;

//     // if (auth) {
//         // const ingredient = await Ingredient.find({})
//         // console.log('authenticated!', auth);
//         return res.send('Hi, from within the ingredients router GET'); 
//     // }
//     // console.log(req.currentUser)
//     // return res.status(403).send('Not authorized get')
// });

createIngredient = (req, res) => {
    const auth = req.currentUser;

    if (auth) {
        // console.log('authenticated!', auth);
        const ingredient = new Ingredient(req.body)
        const savedIngredient = ingredient.save()

        return res.status(201).json(savedIngredient);
    }
    return res.status(403).send('Not authorized post')

}
// ingredientsRouter.post('/ingredients', (req, res) => {
//     // const auth = req.currentUser;

//     // if (auth) {
//     //     console.log('authenticated!', auth);
//         const ingredient = new Ingredient(req.body)
//         const savedIngredient = ingredient.save()

//         return res.status(201).json(savedIngredient);
//     // }
//     // return res.status(403).send('Not authorized post')
// });

removeIngredient = async (req, res) => {
    const auth = req.currentUser;

    if (auth) {

        await Ingredient.findOneAndDelete({ _id: req.params.id }, (err, ingredient) => {
            if (err) {
                return res.status(400).json({ success: false, error: err })
            }

            if (!ingredient) {
                return res
                    .status(404)
                    .json({ success: false, error: `Ingredient not found` })
            }

            return res.status(200).json({ success: true, data: ingredient })
        }).catch(err => console.log(err))

    }
    return res.status(403).send('Not authorized post')

}

module.exports = {
    getIngredients,
    createIngredient,
    removeIngredient
};