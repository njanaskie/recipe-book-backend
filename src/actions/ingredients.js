import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_INGREDIENT
export const addIngredient = (ingredient) => ({
    type: 'ADD_INGREDIENT',
    ingredient
});

export const startAddIngredient = (ingredientData = {}) => {
    return (dispatch) => {
        const {
            name = '',
            category = '',
            price = ''
        } = ingredientData;
        const ingredient = { name, category, price };

        return database.ref('ingredients').push(ingredient).then((ref) => {
            dispatch(addIngredient({
                id: ref.key,
                ...ingredient
            }));
        });
    };
};

// REMOVE_INGREDIENT
export const removeIngredient = ({ id } = {}) => ({
    type: 'REMOVE_INGREDIENT',
    id
});

export const startRemoveIngredient = ({ id } = {}) => {
    return (dispatch, getState) => {
        return database.ref(`ingredients/${id}`).remove().then(() => {
            dispatch(removeIngredient({ id }))
        });
    };
};

// EDIT_INGREDIENT
export const editIngredient = (id, updates) => ({
    type: 'EDIT_INGREDIENT',
    id,
    updates
});

export const startEditIngredient = (id, updates) => {
    return (dispatch, getState) => {
        return database.ref(`ingredients/${id}`).update(updates).then(() => {
            dispatch(editIngredient(id, updates))
        });
    }
}

// SET_INGREDIENTS
export const setIngredients = (ingredients) => ({
    type: 'SET_INGREDIENTS',
    ingredients
});

export const startSetIngredients = () => {
    return (dispatch, getState) => {
        return database.ref('ingredients')
        .once('value')
        .then((snapshot) => {
            const ingredients = [];

            snapshot.forEach((childSnapshot) => {
                ingredients.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setIngredients(ingredients))
        });
    };
};