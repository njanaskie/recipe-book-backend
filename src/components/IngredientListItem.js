import React from 'react'
import { Button, List } from 'semantic-ui-react'
import { useIngredientsContext } from '../../context/ingredients-context'
import database from '../firebase/firebase'

const IngredientListItem = ({ ingredient }) => {
  const { dispatch } = useIngredientsContext()

  const removeIngredient = () => {
    database.collection('ingredients').doc(ingredient.id).delete().then(() => {
      dispatch({ type: 'REMOVE_INGREDIENT', id: ingredient.id })
    })
  }

  return (
    <List.Item>
      <List.Content floated='right'>
        <Button onClick={removeIngredient} icon='delete' size='tiny'/>
      </List.Content>
      <List.Content>
        <List.Header>{ingredient.name}</List.Header>
        <List.Description>{ingredient.category.toUpperCase()}</List.Description>
      </List.Content>
    </List.Item>
  )
}

export { IngredientListItem as default }