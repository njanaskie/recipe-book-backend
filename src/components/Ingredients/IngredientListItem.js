import React, { useContext, useState, useEffect } from 'react'
import FirebaseContext from '../../../context/firebase-context'
import IngredientsContext from '../../../context/ingredients-context'
import PantryContext from '../../../context/pantry-context'
import database, { firebase } from '../../firebase/firebase'

const IngredientListItem = ({ ingredient } ) => {
  const { user } = useContext(FirebaseContext)
  const { dispatch } = useContext(IngredientsContext)
  const { pantryIngredients, pantryDispatch } = useContext(PantryContext)

  const pathname = window.location.pathname

  const removeIngredient = () => {
    database.collection('ingredients').doc(ingredient.id).delete().then(() => {
      dispatch({ type: 'REMOVE_INGREDIENT', id: ingredient.id })
    })
  }

  const addPantryIngredient = () => {
    const uid = user.uid

    database.collection('users').doc(uid).collection('pantry').add(ingredient).then(() => {
      pantryDispatch(({ type: 'ADD_PANTRY_INGREDIENT', pantryIngredient: {...ingredient} }))
    })
  }

  const removePantryIngredient = () => {
    const uid = user.uid

    database.collection('users').doc(uid).collection('pantry').doc(ingredient.id).delete().then(() => {
      dispatch({ type: 'REMOVE_PANTRY_INGREDIENT', id: ingredient.id })
    })
  }

  return (
    <div className='list-item' >

      <h3>{ingredient.name}</h3>
      <p>{ingredient.category}</p>
      <p>{ingredient.price}</p>
      {pathname === '/pantry' ?
        <div>
          {ingredient.isPantry &&
            <span>P</span>
          }
          <button onClick={addPantryIngredient}>
            Add to Pantry
          </button> 
          <button onClick={removePantryIngredient}>
            Remove from Pantry
          </button>
        </div>
        :
        <button onClick={removeIngredient}>Remove</button>}
    </div>
  )
}

export { IngredientListItem as default }