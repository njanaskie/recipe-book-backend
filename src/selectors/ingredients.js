
export default (ingredients) => {
    return ingredients.filter(ingredient => ingredient.id !== undefined
    )
    .sort((a, b) => {
        if (a.category === b.category) {
            return b.category - a.category
        }
         return a.category > b.category ? 1 : -1
    })
}