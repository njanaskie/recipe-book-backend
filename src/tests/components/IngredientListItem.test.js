import React from 'react'
import IngredientsContext from '../../../context/ingredients-context'
import { shallow } from 'enzyme'
import IngredientListItem from '../../components/Ingredients/IngredientListItem'
import ingredients from '../fixtures/ingredients'

test('should render IngredientListItem with ingredient',() => {
    const wrapper = shallow(
        <IngredientsContext.Provider value={ingredients}>
            <IngredientListItem {...ingredients[2]}/>
        </IngredientsContext.Provider>)
    expect(wrapper).toMatchSnapshot()
});