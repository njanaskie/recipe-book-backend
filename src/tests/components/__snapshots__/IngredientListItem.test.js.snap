// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render IngredientListItem with ingredient 1`] = `
<IngredientListItem
  category="Dairy"
  id="3"
  name="Yoghurt"
  price={2}
/>
`;
